src/model/element-base.ts

```diff
@@ -95,7 +131,180 @@ export abstract class ElementBase implements Feature {
    * single list. A mixin can be applied more than once, each time its
    * members override those before it in the prototype chain.
    */
-  mixins: Reference[] = [];
+  readonly mixins: ImmutableArray<Reference>;
+
+  constructor(init: ElementBaseInit, document: Document) {
+    const {
+      description,
+      summary,
+      demos,
+      events,
+      sourceRange,
+      properties,
+      attributes,
+      methods,
+      astNode,
+      warnings,
+      slots,
+      privacy,
+      jsdoc,
+      superClass,
+      mixins,
+    } = init;
+    this.description = description;
+    this.summary = summary;
+    this.demos = demos ? Array.from(demos) : [];
+    this.sourceRange = sourceRange;
+    this.astNode = astNode;
+    this.warnings = warnings ? Array.from(warnings) : [];
+    this.slots = slots ? Array.from(slots) : [];
+    this.privacy = privacy;
+    this.jsdoc = jsdoc;
+    this.superClass = superClass ? superClass.resolve(document) : undefined;
+    this.mixins = (mixins || []).map((m) => m.resolve(document));
+
+    const prototypeChain = this._getPrototypeChain(document, init);
+    for (const superClass of prototypeChain) {
+      this.inheritFrom(superClass);
+    }
+
+    this._overwriteInherited(
+        this.properties, properties || [], undefined, true);
+    this._overwriteInherited(
+        this.attributes, attributes || [], undefined, true);
+    this._overwriteInherited(this.methods, methods || [], undefined, true);
+    this._overwriteInherited(this.events, events || [], undefined, true);
+
+    for (const method of this.methods) {
```

Comment:
we can't mutate `methods`, only `this.methods`. this is awkward and should be removed entirely though, added a comment.